mitmproxy/proxy/layers/quic.py

```diff
@@ -0,0 +1,1052 @@
+        input_buf: QuicBuffer,
+        initial_buf: QuicBuffer,
+        handshake_buf: QuicBuffer,
+        onertt_buf: QuicBuffer,
+    ) -> None:
+        assert input_buf.pull_uint8() == HandshakeType.CLIENT_HELLO
+        length = 0
+        for b in input_buf.pull_bytes(3):
+            length = (length << 8) | b
+        offset = input_buf.tell()
+        raise QuicClientHello(input_buf.data_slice(offset, offset + length))
+
+    def initialize_replacement(peer_cid: bytes) -> None:
+        try:
+            return _initialize(peer_cid)
+        finally:
+            quic.tls._server_handle_hello = server_handle_hello_replacement
+
+    quic._initialize = initialize_replacement
+    try:
+        quic.receive_datagram(data, ("0.0.0.0", 0), now=0)
+    except QuicClientHello as hello:
+        try:
+            return ClientHello(hello.data)
+        except EOFError as e:
+            raise ValueError("Invalid ClientHello data.") from e
+    except QuicConnectionError as e:
+        raise ValueError(e.reason_phrase) from e
+    raise ValueError("No ClientHello returned.")
+
+
+class QuicStream:
+    flow: tcp.TCPFlow | None
+    stream_id: int
+
+    def __init__(self, context: context.Context, stream_id: int, ignore: bool) -> None:
+        if ignore:
+            self.flow = None
+        else:
+            self.flow = tcp.TCPFlow(context.client, context.server, live=True)
+        self.stream_id = stream_id
+        is_unidirectional = stream_is_unidirectional(stream_id)
+        from_client = stream_is_client_initiated(stream_id)
+        self._ended_client = is_unidirectional and not from_client
+        self._ended_server = is_unidirectional and from_client
+
+    def has_ended(self, client: bool) -> bool:
+        return self._ended_client if client else self._ended_server
+
+    def mark_ended(self, client: bool, err: str | None = None) -> layer.CommandGenerator[None]:
+        # ensure we actually change the ended state
+        if client:
+            assert not self._ended_client
+            self._ended_client = True
+        else:
+            assert not self._ended_server
+            self._ended_server = True
+
+        # we're done if the stream is ignored
+        if self.flow is None:
+            return
+
+        # set and report the first error
+        if err is not None and self.flow.error is None:
+            self.flow.error = mitm_flow.Error(err)
+            yield TcpErrorHook(self.flow)
+
+        # report error-free endings and always clear the live flag
+        if self._ended_client and self._ended_server:
+            if self.flow.error is None:
+                yield TcpEndHook(self.flow)
+            self.flow.live = False
+
+
+class QuicStreamLayer(layer.Layer):
+    """
+    Layer on top of `ClientQuicLayer` and `ServerQuicLayer`, that simply relays all QUIC streams and datagrams.
+    This layer is chosen by the default NextLayer addon if ALPN yields no known protocol.
```

Comment:
Clarification question: This would also work perfectly fine to proxy HTTP/3 streams, correct?

Reply:
Yes, I've actually tested that in the previous implementation (before multi-server). This layer is basically TCPLayer/UDPLayer for QUIC.

Reply:
Awesome!